### PR TITLE
🐛 Fix: Packages dependencies

### PR DESCRIPTION
**Pull request by killian31**
# What does this PR do?

This pull request makes the package compatible with the latest versions of packages, since there has been breaking changes in some of the dependencies since 3 years ago 😃.

# List of the fixes

## Dataloader
- To get rid of this warning,

  ```raw
  The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=7` in the `DataLoader` to improve performance.
  ```
  
  changed default of `num_workers` from 0 to 7

- To get rid of this warning,

  ```raw
  Consider setting `persistent_workers=True` in 'train_dataloader' to speed up the dataloader worker initialization.
  ```
  
  added `persistent_workers=True` in the DataLoader initialisation call.

### Trainer
The `progress_bar_refresh_rate` and `gpus` argument are deprecated, replaced the former with `enable_progress_bar=True`.

### Metrics
Lightning metrics have been moved to their own package, torchmetrics, as stated in this [issue](https://github.com/Lightning-AI/pytorch-lightning/issues/2207#issuecomment-970647682). 

Replaced `self.accuracy = pl.metrics.Accuracy()` with `self.accuracy = torchmetrics.Accuracy(task="multiclass", num_classes=num_classes)` and added a `num_classes` argument to the model class since it is required by `Accuracy`.

### Method `on_train_epoch_end()`
The method `training_epoch_end(training_step_outputs)` is deprecated and needs to be changed for `on_train_epoch_end()`.  The `training_step_outputs` list is initialised in the constructor to be accessed in `on_train_epoch_end()`

Thank you for your work, I hope those fixes can benefit the research community!

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -16,14 +16,14 @@ include = [
...
-pytorch-lightning = "^1.1.2"
-sklearn = "^0.0"
-torchvision = "^0.8.2"
-importlib_resources = "^5.0.0"
-tqdm = "^4.56.0"
+numpy = "^1.26.3"
+torch = "^2.1.2"
+matplotlib = "^3.8.2"
+pytorch-lightning = "^2.1.3"
+sickit_learn = "^1.4.0"
```
Comment (logancyang) [inline pyproject.toml:23]: Ah missed it, this is wrong. You should also run poetry lock after you update deps here

Comment (killian31) [reply pyproject.toml:23]: Yep I know but for some reason poetry won't run on my computer.. I'll try to fix it.
